src/schedule/ll_schedule.c

```diff
@@ -63,18 +63,29 @@ struct ll_queue_shared_context {
 
 static struct ll_queue_shared_context *ll_shared_ctx;
 
+static inline uint64_t ll_get_timer(struct ll_schedule_data *queue)
+{
+	return queue->ts->timer_get(&queue->ts->timer);
+}
+
 /* calculate next timeout */
 static inline uint64_t queue_calc_next_timeout(struct ll_schedule_data *queue,
 					       uint64_t start)
 {
-	return queue->ticks_per_msec * queue->timeout / 1000 + start;
-}
+	uint64_t ticks = queue->ticks_per_msec * queue->timeout / 1000 + start;
 
-static inline uint64_t ll_get_timer(struct ll_schedule_data *queue)
-{
-	return queue->ts->timer_get(&queue->ts->timer);
+	/* Did queued tasks took more time than scheduled period?
+	 * NOTE: We also take into account ticks needed to finish
+	 * rescheduling procedure ().
+	 */
+	if (ticks < ll_get_timer(queue) + TICKS_NEEDED_TO_FINISH_RESCHEDULE)
+		ticks = ll_get_timer(queue) +
```

Comment:
We clearly miss such error handling, but your solution is not perfect.
Setting ticks to current time will desynchronize the whole queue, because you don't take into account start, which is ll_shared_ctx->last_tick. We should set ticks to minimum value > ll_get_timer(queue) + TICKS_NEEDED_TO_FINISH_RESCHEDULE, which is also multiplication of originally calculated value.

Reply:
`ll_shared_ctx->last_tick` is past time so why would we still take it into account? I think the only reasonable reference to us is current timer value.

Reply:
@tlauda, this newly calculated timeout is assumed to be minimum delayed value.

Reply:
`ll_shared_ctx->last_tick` is set to returned ticks, so you're right. Missed that one.